### ISSUE TITLE
Idea: Make door keys interchangeable

### ISSUE THREAD

**Issue by HomebrewHomunculus** [HomebrewHomunculus/BrogueLite#23]
Currently, in Brogue, keys are generated in special keyholder rooms, and there is always exactly one key per vault door. Also, as far as I know, the key always generates on the same depth as the door it's for is. Keys cannot be used on other floors, and when there are two vaults on the same floor, the keys are unique, i.e. key 1 is only accepted by vault 1. Key 2 may be placed in vault 1, in which case the rooms have a definite ordering.

We could make keys non-unique and interchangeable, so that any key from any floor could fit any vault door.

This would loosen Brogue's straitjacket on the player in several ways:
* Players could use a spare key to avoid having to do a keyholder room that looks dangerous or time-consuming. (Especially if there are more keys than doors, or if keys can appear several depths before their doors.)
* Players could scout the items in the vault by peeking through windows, and decide which vault they want to get (first).
* Clairvoyance would gain an additional use case in the scouting of vaults.
* Players could decide to save the key for later if they don't like the look of the items. This adds a layer of strategy and resource management to vaults, which currently aren't very integrated with the rest of the game: usually getting access to the vault requires manipulating some gameplay mechanics, but once the vault is accessed, the gameplay ends.
* Since vaults are now accessible without their keyholders, the rare cases where a vault is "broken" and readily accessed are no longer serious bugs, but just an interesting part of the game's systems. Things like allowing teleporting into a vault could be perfectly fine. (Of course, the player would need a key or second teleport to escape the vault...)

And that's all without changing level generation whatsoever. (If we consider changes there, then fun things like extra keys carried by monsters, in secret rooms, etc. become possible.)

Keys should stack in the inventory, so that they don't completely fill it up. Or should they? Should lugging many keys around be a tradeoff for inventory space? Perhaps that would be a good way to mitigate the problem of hoarding ("there might be a really good vault later, so I'll save my key!")

(Also, keys could make for an improvised throwing weapon, perhaps.)

Possible pitfalls:
* Player confusion. 
* Accidentally using a key on vault 2, which means that the key in vault 1 is missed.

**Comment by HomebrewHomunculus** [comment]
This also brings up other avenues that could be quite interesting. In issue #20 and https://github.com/tmewett/BrogueCE/issues/342 we've been throwing around some ideas for reworking scores, by removing gold or making it more meaningful, and/or allowing players to choose between score and power-ups. I brought up the idea of having gold (or any equivalent score item) as an option in vaults, so that the player would have to choose between getting better equipment and upping their score. But here's a thought: **what if keys were the scoring item**, rather than gold? Then the choice would be moved to an earlier point. This would remove gold as essentially a redundant item. Keys are a critical component of Brogue, but now they would be overloaded with multiple uses, in a quite elegant way.

Furthermore, the number of dungeon features that accept keys (whether vault doors or other, new features) in a single dungeon is currently dependent on the number of keys. That is, if the dungeon has _N_ vaults, then _N_ keys are generated. However, with fungible keys, there can be _N_ vaults and _M_ keys, and the amounts are independent of each other. The interesting possibility that this creates is that by adjusting _M_, we can adjust the difficulty level of the game, possibly without even changing level generation. The same dungeon with an allotment of 10 keys is much harder than with 20 keys, even if the vaults are identical.

This gives us the possibility of increasing the amount of resources available in the game without really changing its internals as long as the resources consume a key AND the number of keys is kept constant. And, therefore, the game can also be made easier than it is now by upping the number of keys.

Take, for example, a single dungeon in Brogue CE in its current form. Now add to that dungeon 10 places where the player may trade a vault key for additional food, but don't change the number of keys. Has the dungeon become easier? There is more food available if the player wants it, but most players probably don't feel like the food clock is a big threat, just an incentive to move around efficiently, so they wouldn't need to take the extra food often. On the other hand, a player who does struggle with the food clock can trade keys for food - but it's a tradeoff, missing out on powerful items. But there are players who find that the food clock is an obstacle for them getting into Brogue, so giving them this option could help them without increasing the food allotment for all players.

Now consider all the other resources than food which could be exchanged for keys. Extra potions of Strength, perhaps? It's an extreme example, but not beyond the realm of possibility. Some may say that it would make the game far too easy - and perhaps for them it would, if they were disregarding score completely, but it would also open up a new challenge mode: if the game's so easy for you, then see how few keys you can beat it with! Maybe novice players get 30 keys and that will make the game much easier for them to ascend in than in standard Brogue. Maybe players for whom standard Brogue is the sweet spot will only use 20 keys. And maybe hardcore players will try to brave the dungeon with 10 keys or even less. Thus, their scores would be 0/10/20+ (times whatever score value each key is given).

On the other hand, even without any scoring changes, a difficulty selection could easily be created: "easy mode" is starting the game with lots of extra keys that can be used to avoid obstacles or acquire resources.

**Comment by HomebrewHomunculus** [comment]
A less radical variant of this idea is allowing enchantment to turn a regular key into a skeleton key that fits any door.

Another possibility is to have vaults freely entered, but have the altar cages require a key to open. This means that it's possible to take two items from one reward room if you have two keys.

**Comment by tmewett** [comment]
I like this idea! My only concern is entering nested vaults in the wrong order, if it's not possible to know which vault to enter first.

Additionally, in the thread in CE you ask:

> By the way, is there any reason why the "item library" type vault, I mean the one with altars and falling cages, couldn't have the keys on the cages instead?

A factor is that the items are only identified when entering the vaults. This means you can't peek in through windows and evaluate vaults without committing to them. This might need changing to support some of the plays in your post.. unless I'm mistaken

**Comment by HomebrewHomunculus** [comment]
> My only concern is entering nested vaults in the wrong order, if it's not possible to know which vault to enter first.

Yes, that's what I was thinking (and tried to express above, but didn't know nested vaults was the term I was looking for - thanks :) 

How I've tried to solve this in Brogue Lite for now is to give the player 1-2 extra keys at the start of the game - I wanted to avoid having to change level generation. It's quite generous, sure, but on the other hand it might give the player some additional interesting choices of when to spend their extra keys. They could save the keys for skipping a keyholder of their choice, for example, but only in a limited fashion. 

(This is how it is currently in the master branch, if you wanna check it out. But I had to do a bit of hacking with the key tags to get it to only apply to door keys, and I haven't tested yet if it breaks crystal orbs.)

![image](https://user-images.githubusercontent.com/80920451/123039314-94ffaa00-d3fa-11eb-846b-d263ec739c64.png)

> A factor is that the items are only identified when entering the vaults. This means you can't peek in through windows and evaluate vaults without committing to them. This might need changing to support some of the plays in your post.. unless I'm mistaken

This is not a problem in Brogue Lite, where all items start out as identified :) 

I think I will not try to implement any key scoring/trading for now, but I do want to get out a slightly more experimental version of Lite with extra door keys + scoring for depth/less score for gold (and lower score numbers in general). Less pressure to get everything right the first time when it's not the mainline...
